server/src/main/java/org/apache/lucene/index/BaseSingleStarTreeBuilder.java

```diff
@@ -0,0 +1,660 @@
+                }
+            }
+            return aggregatedStarTreeDocument;
+        }
+    }
+
+    /**
+     * Merges a star-tree document (aggregated) into the aggregated document.
+     *
+     * <p>Will create a new aggregated starTreeDocument if the current one is {@code null}.
+     *
+     * @param aggregatedStarTreeDocument Aggregated star-tree document
+     * @param starTreeStarTreeDocument   Star-tree document
+     * @return Merged star-tree document
+     */
+    public StarTreeDocument aggregateStarTreeDocument(
+        StarTreeDocument aggregatedStarTreeDocument,
+        StarTreeDocument starTreeStarTreeDocument
+    ) {
+        // aggregate the documents
+        if (aggregatedStarTreeDocument == null) {
+            long[] dimensions = Arrays.copyOf(starTreeStarTreeDocument.dimensions, numDimensions);
+            Object[] metrics = new Object[numMetrics];
+            for (int i = 0; i < numMetrics; i++) {
+                try {
+                    metrics[i] = valueAggregators[i].cloneAggregatedValue(starTreeStarTreeDocument.metrics[i]);
+                } catch (IllegalArgumentException | NullPointerException e) {
+                    logger.error("Cannot clone aggregated value", e);
+                    throw new IllegalArgumentException("Cannot clone aggregated value [" + starTreeStarTreeDocument.metrics[i] + "]");
+                }
+            }
+            return new StarTreeDocument(dimensions, metrics);
+        } else {
+            for (int i = 0; i < numMetrics; i++) {
+                try {
+                    aggregatedStarTreeDocument.metrics[i] = valueAggregators[i].applyAggregatedValue(
+                        starTreeStarTreeDocument.metrics[i],
+                        aggregatedStarTreeDocument.metrics[i]
+                    );
+                } catch (IllegalArgumentException | NullPointerException e) {
+                    logger.error("Cannot apply aggregated value", e);
+                    throw new IllegalArgumentException("Cannot apply aggregated value [" + starTreeStarTreeDocument.metrics[i] + "]");
+                }
+            }
+            return aggregatedStarTreeDocument;
+        }
+    }
+
+    // TODO: This will be taken care in off heap implementation for merges
+    // public abstract void build(List<StarTreeValues> starTreeValues) throws IOException;
+
+    public void build() throws IOException {
+        long startTime = System.currentTimeMillis();
+        logger.info("Tree of Aggregations build is a go with config {}", starTreeField);
+
+        Iterator<StarTreeDocument> starTreeDocumentIterator = processSegmentStarTreeDocuments(totalDocs);
+        logger.info("Sorting and aggregating star-tree in ms : {}", (System.currentTimeMillis() - startTime));
+        build(starTreeDocumentIterator);
+        logger.info("Finished Building TOA in ms : {}", (System.currentTimeMillis() - startTime));
```

Comment:
nit : TOA ?

Reply:
Tree of Aggregations (missed removing this, was playing around with names)